mempool: Add mempool tracepoints

virtu opened the pull request:
This PR adds multiple mempool tracepoints.

| tracepoint  | description |
| ------------- | ------------- |
| `mempool:added`  | Is called when a transaction enters the mempool  |
| `mempool:removed`  | ... when a transaction is removed from the mempool |
| `mempool:replaced`  | ... when a transaction is replaced in the mempool |
| `mempool:rejected`  | ... when a transaction is rejected from entering the mempool |

The tracepoints are further documented in `docs/tracing.md`. Usage is demonstrated in the example script `contrib/tracing/mempool_monitor.py`. Interface tests are provided in `test/functional/interface_usdt_mempool.py`.

The rationale for passing the removal reason as a string instead of numerically is that the benefits of not having to maintain a redundant enum-string mapping seem to outweigh the small cost of string generation. The reject reason is passed as string as well, although in this instance the string does not have to be generated but is readily available.

virtu committed:
Add mempool tracepoints

Tracepoints for added, removed, replaced, and rejected transactions.

The removal reason is passed as string instead of a numeric value, since
the benefits of not having to maintain a redundant enum-string mapping
seem to outweigh the small cost of string generation.  The reject reason
is passed as string as well, although here the string does not have to
be generated but is readily available.

So far, tracepoint PRs typically included two demo scripts: a naive
bpftrace script to show raw tracepoint data and a bcc script for a more
refined view. However, as some of the ongoing changes to bpftrace
introduce a certain degree of unreliability (running some of the
existing bpftrace scripts was not possible with standard kernels and
bpftrace packages on latest stable Ubuntu, Debian, and NixOS), this PR
includes only a single bcc script that fuses the functionality of former
bpftrace and bcc scripts.

0xB10C commented:
Concept ACK

jamesob commented on doc/tracing.md:
```diff
@@ -211,6 +211,58 @@ Arguments passed:
+Arguments passed:
+1. Transaction ID (hash) as `pointer to unsigned chars` (i.e. 32 bytes in little-endian)
+2. Transaction virtual size as `uint64`
+3. Transaction fee as `int64`
```
Would it make sense to convey information about the peer (even if just ID) as well? One of the things that I think is potentially valuable e.g. [in bmon](https://github.com/chaincodelabs/bmon/blob/master/bmon/logparse.py#L245) is tracking which peers fed you which transactions.

virtu replied:
I think it makes sense.

However, peer information is no longer available in functions that handle low-level mempool operations, so I propose adding a separate `mempool:accepted` tracepoint for this in `PeerManagerImpl::ProcessMessage` [here](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L4042) (which is right where the debug log entries for transactions accepted into the memory pool are created).

The rationale for adding a separate `mempool:accepted` tracepoint instead of moving the existing `mempool:added` one from `CTxMemPool::addUnchecked` is to ensure consistency of added and removed transactions.

Details: The tracepoint call in `ProcessMessage` would depend on the success of `ProcessTransaction` [here](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L4025). There's an edge case where a transaction added to the mempool in `MemPoolAccept::Finalize` [here](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1074) is immediately removed again seven lines below during the call to `LimitMempoolSize` and `MemPoolAccept::Finalize` will return `state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "mempool full")`. As expected, in this case `ProcessMessage` will not call the tracepoint (or create a debug log entry). `LimitMempoolSize`, however, will at some point call `CTxMemPool::removeUnchecked`, which calls the `mempool:removed` tracepoint, resulting in tracepoint data containing a "transaction removed" event without a preceding "transaction added" event.

Please note that this problem does not arise due to poor placement of the "mempool:removed" tracepoint [here](https://github.com/virtu/bitcoin/blob/add_mempool_tracepoints/src/txmempool.cpp#L481-L488). The tracepoint is positioned right after `GetMainSignals().TransactionRemovedFromMempool`, which leads me to believe that the aforementioned consistency issue currently affects zeromq and the debug log.

0xB10C replied:
Note that adding transactions we receive over the network is only one option out of several to add something to our mempool. Can also add transactions via RPC, `loadMempool()`, wallet, etc..   

I agree that it could be useful to have information about peers that send us transactions we end up accepting.  Could call a tracepoint something like `net:tx_accepted`. Not sure if that's in-scope for this PR though.

dergoegge replied:
Couldn't this be done by using both `net:inbound_message` and `mempool:added` at the same time? That is, correlate the txids from both tracepoints to deduce which peer sent the tx.

jamesob reviewed (commented):
Cool! concept ACK

glozow commented on contrib/tracing/mempool_monitor.py:
```diff
@@ -0,0 +1,361 @@
+import sys
+from datetime import datetime, timezone
+
+from bcc import BPF, USDT
```
```
test/functional/interface_usdt_mempool.py:14: error: Cannot find implementation or library stub for module named "bcc"  [import]
test/functional/interface_usdt_mempool.py:14: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 256 source files)
^---- failure generated from lint-python.py
```


I think you may want to do this?
```suggestion
# Test will be skipped if we don't have bcc installed
try:
    from bcc import BPF, USDT # type: ignore[import]
except ImportError:
    pass
```

0xB10C replied:
yes, that's the trick we're using in the other tracepoint tests too

virtu replied:
Thanks.

glozow reviewed (commented):
Concept ACK, thanks for working on this!

dergoegge commented:
Concept ACK

0xB10C commented:
With [#26593: tracing: Only prepare tracepoint arguments when actually tracing](https://github.com/bitcoin/bitcoin/pull/26593), we could also pass the serialized transactions as arguments for the mempool tracepoints. I think this makes these mempool tracepoints even more interesting as we otherwise only have the txids of removed/rejected/replaced transactions and wouldn't need to pair them with e.g. ZMQ `rawtx` data in an extra step.  As demonstration, see the last three commits of [2022-11-virtu-add_mempool_tracepoints-with-rawtx](https://github.com/0xB10C/bitcoin/commits/2022-11-virtu-add_mempool_tracepoints-with-rawtx). I'd be interested in hearing from people that left a Concept ACK here and @virtu if this extension to this PR that makes sense (once we have more Concept ACKs for #26593).

I'm currently working on monitoring for fullrbf replacements as seen by my node with the above branch (if the replaced transaction didn't signal optin-RBF, it should be a fullrbf replacement). 

(also sorry for the "0xB10C added a commit to 0xB10C/bitcoin that referenced this pull request" spam. Forgot that this would appear here for every force push...)

virtu commented:
> One note should removal reason 'BLOCK' be logged? It's the one exclusion from even ZMQ because of the resulting data dump. With eBPF it's the most difficult to log because it will overflow the eBPF ring buffer every time.

Block removal was deliberately included because it makes data collection much easier (inferring removal times from block arrival time involves a certain degree of processing; also, timestamps could be wrong if there was a block reorg and the original block is lost) and the overhead of calling the tracepoint pales in comparison to the rest of the code. But that's not set in stone.

Check the `mempool_monitor.py`'s [code](https://github.com/bitcoin/bitcoin/pull/26531/files#diff-f0c928fac39782cb8cefde1f6345d08195e0c359a68aa53d2ed6af6ca45208cbR152-R158) for a possible take on avoiding overflowing the ring buffer.

0xB10C commented on contrib/tracing/README.md:
```diff
@@ -286,3 +286,49 @@ Spent   a05880b8c77971ed0b9f73062c7c4cdb0ff3856ab14cbf8bc481ed571cd34b83:1
+ └───────────────────────────────────┘  └─────────────────────────────────────┘
+
+ ┌─Event log───────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+ │ 17:09:56Z event=added, txid=0a2f2ef84d77a13162f26bb7a544efbab7f09d0b8a965b6f00e85bbd64cd8152, vsize=183, fee=4968   │
+ │ 17:09:56Z event=removed, txid=0356df4450c76d2474eab13fab15bfab843e31617edf9ce88d5d5bea5b19bf3e, reason=replaced, vsi│
+ │ ze=110, fee=16655                                                                                                   │
+ │ 17:09:56Z event=removed, txid=653db4f9af156626a7359c18f9cb095009583497ef606d7a9d2f093c63d392da, reason=replaced, vsi│
+ │ ze=110, fee=4037                                                                                                    │
+ │ 17:09:56Z event=rejected, txid=1cd12698577446904e8654372421e1ab5b82bdb8faee3f11f34ed758c17ae85a, reason=bad-txns-inp│
+ │ uts-missingorspent, peer_id=2568, peer_addr=158.69.116.169:8333                                                     │
+ │ 17:09:56Z event=replaced, replacement_txid=db3c91d7dfff65ea5b85ca58bc3e9a26556eb6506ccb04a573f920d297060218, replace│
+ │ ment_vsize=110, replacement_fee=23343, replaced_txid=0356df4450c76d2474eab13fab15bfab843e31617edf9ce88d5d5bea5b19bf3│
+ │ e, replaced_vsize=110, replaced_fee=16655                                                                           │
+ │ 17:09:56Z event=replaced, replacement_txid=9b304b9357df5e2cfcbbbead7ffa67c0efc3c599aac9093362b398478d46a587, replace│
+ │ ment_vsize=110, replacement_fee=4268, replaced_txid=653db4f9af156626a7359c18f9cb095009583497ef606d7a9d2f093c63d392da│
+ │ , replaced_vsize=110, replaced_fee=4037                                                                             │
```
Played around with this a bit. Super cool!

nit: I think these could be a bit more human readable instead of just printing the raw values. Examples:
- added 0356df4450c76d2474eab13fab15bfab843e31617edf9ce88d5d5bea5b19bf3e with 27.14 sat/vbyte (4968 sat in fees at 183 vbyte)
- replaced 0356df4450c76d2474eab13fab15bfab843e31617edf9ce88d5d5bea5b19bf3 (110 vbyte, 16655 sat,  151.40 sat/vbyte) with db3c91d7dfff65ea5b85ca58bc3e9a26556eb6506ccb04a573f920d297060218 (110 vbyte, 23343 sat, 212.20 sat/vbyte)
- rejected 1cd12698577446904e8654372421e1ab5b82bdb8faee3f11f34ed758c17ae85a from peer 2568 (158.69.116.169:8333): bad-txns-inputs-missingorspent

0xB10C commented on doc/tracing.md:
```diff
@@ -211,6 +211,58 @@ Arguments passed:
+Is called when a transaction is removed from the node's mempool. Passes information
+about the transaction.
+
+Arguments passed:
+1. Transaction ID (hash) as `pointer to unsigned chars` (i.e. 32 bytes in little-endian)
+2. Removal reason as `pointer to C-style String` (max. length 9 characters)
+3. Transaction virtual size as `uint64`
+4. Transaction fee as `int64`
+
```
probably makes sense to also pass the time the transaction was initially added to the mempool here. `CTxMempoolEntry` has a `GetTime()` method.

achow101 replied:
It would probably make sense to include the time for all of the tracepoints added in this PR?

0xB10C replied:
Not necessarily. Might be redundant. When a transaction is added and you pass the tracepoint arguments _directly_ to a tracing script (can't think of a reason why you would want to store it in a BPF map first for a while), then the time is very close to `now()`. Similar for rejected transactions and for the replacement transaction during RBF (the time when the replaced transaction entered is passed).

virtu replied:
In reply to the original comment: oversight on my part; will push a fix.

In reply to adding event timestamps for all tracepoints: in omitting the actual event timestamps as tracepoint parameters, I followed the precedent set by other tracepoints, where, as laid out by @0xB10C, event timestamps are generated inside tracing scripts.

virtu commented:
- addressed feedback on [argument order of mempool:replaced](https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1036914820) with [2335347](https://github.com/bitcoin/bitcoin/pull/26531/commits/2335347474e87c707008e6ac8b6632a0af751aea): replaced transaction details now passed before replacement transaction details
- addressed feedback on [adding replaced transaction's entry time](https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1045624415) with [bfe4fe4](https://github.com/bitcoin/bitcoin/pull/26531/commits/bfe4fe444345e7d3215d93754de8d1b86e9753cc): added replaced transaction's entry time to `mempool:replaced` tracepoint
- addressed feedback on [making demo script output more human readable](https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1039635761) with [869a749](https://github.com/bitcoin/bitcoin/pull/26531/commits/869a7492df02e85383f55b9bac96a24dadf6d72f): added replaced transaction's entry time to `mempool:replaced` tracepoint
- feedback on [adding serialized tx to tracepoints once they're semaphore-gated](https://github.com/bitcoin/bitcoin/pull/26531#issuecomment-1333832906): Strongly in favor of this, @0xB10C! If  #26593 is merged before this PR, I will make the necessary changes.

glozow reviewed (commented):
Tested the monitor - this is awesome! A few small questions

virtu commented on test/functional/interface_usdt_mempool.py:
```diff
@@ -0,0 +1,360 @@
+        self.num_nodes = 1
+        self.extra_args = [
+            [
+                "-acceptnonstdtxn=1",
```
I guess I was too occupied with trying to understand the functional test framework to figure out this solution. 
[d890e9a](https://github.com/bitcoin/bitcoin/pull/26531/commits/d890e9ae605de4f8317cd2bdc475a8793dfe6aae) removes `acceptnonstdtxn=1`.

0xB10C replied:
(I think this is a leftover from when you used ToManySigOps to create an invalid transaction in an earlier version of this PR)

0xB10C commented on src/net_processing.cpp:
```diff
@@ -4161,6 +4161,12 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             LogPrint(BCLog::MEMPOOLREJ, "%s from peer=%d was not accepted: %s\n", tx.GetHash().ToString(),
                 pfrom.GetId(),
                 state.ToString());
+            TRACE4(mempool, rejected,
```
I initially suggested to @virtu to include the peer_id (and thus place it in net_processing.cpp). But based on the 'mempool' scope of this PR might be better to place this tracepoint somewhere transaction source (p2p/wallet/rpc,..) agnostic.

virtu replied:
My reasoning was to follow the precedent set by the debug log, which only records rejected transactions received by peers. I assumed this would be the most interesting case, especially since it allowed the collection of peer data (as suggested by @0xB10C). At the time this approach seemed pragmatic, but now it appears to clash with a clean design.

Let me know what you think of the following change:
1. Remove all networking-related information from the `mempool:rejected` tracepoint to narrow the scope the mempool tracepoints to mempool data, making for a clean interface.
2. Call the `mempool:rejected` tracepoint after [this loop](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1424-L1425) if `test_accept` is `false`.

0xB10C replied:
Sounds good. Might also make sense to call the tracepoint for the `test_accept == false` case. Maybe even pass `test_accept` as argument to allow filtering in e.g. BPF or downstream tools.

virtu replied:
Implemented in [069af24](https://github.com/bitcoin/bitcoin/commit/069af2404c0b71da2857766752d2296e621ca754).

This also simplified the functional tests, since validation no longer requires determining node address and port.

To test, for example, that transactions received via RPC are covered, one can try spending an already-spent tx via `bitcoin-cli`. Example (courtesy of [stackoverflow](https://stackoverflow.com/questions/38493893/heres-how-to-send-raw-transaction-btc-using-bitcoin-cli-command)):

```
bitcoin-cli sendrawtransaction 0100000001e34ac1e2baac09c366fce1c2245536bda8f7db0f6685862aecf53ebd69f9a89c000000006a47304402203e8a16522da80cef66bacfbc0c800c6d52c4a26d1d86a54e0a1b76d661f020c9022010397f00149f2a8fb2bc5bca52f2d7a7f87e3897a273ef54b277e4af52051a06012103c9700559f690c4a9182faa8bed88ad8a0c563777ac1d3f00fd44ea6c71dc5127ffffffff02a0252600000000001976a914d90d36e98f62968d2bc9bbd68107564a156a9bcf88ac50622500000000001976a91407bdb518fa2e6089fd810235cf1100c9c13d1fd288ac00000000
```

arzanilorestan commented:
Hi if you wish, I can send a lot of useful and important news from the true
value of wealth and the important changes of money and assets of the world
markets and all the digital currency information that has been a secret
from all the first to humanity. If you really want to understand them, let
me know to send them all.

در تاریخ جمعه ۱۸ نوامبر ۲۰۲۲،‏ ۲۱:۴۱ Gloria Zhao ***@***.***>
نوشت:

> ***@***.**** commented on this pull request.
>
> Concept ACK, thanks for working on this!
> ------------------------------
>
> In contrib/tracing/mempool_monitor.py
> <https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1026742783>:
>
> > @@ -0,0 +1,361 @@
>
> +#!/usr/bin/env python3
>
> +# Copyright (c) 2022 The Bitcoin Core developers
>
> +# Distributed under the MIT software license, see the accompanying
>
> +# file COPYING or http://www.opensource.org/licenses/mit-license.php.
>
> +
>
> +""" Example logging Bitcoin Core mempool events using the mempool:added,
>
> +    mempool:removed, mempool:replaced, and mempool:rejected tracepoints. """
>
> +
>
> +import curses
>
> +import sys
>
> +from datetime import datetime, timezone
>
> +
>
> +from bcc import BPF, USDT
>
>
> test/functional/interface_usdt_mempool.py:14: error: Cannot find implementation or library stub for module named "bcc"  [import]
>
> test/functional/interface_usdt_mempool.py:14: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
>
> Found 1 error in 1 file (checked 256 source files)
>
> ^---- failure generated from lint-python.py
>
>
> I think you may want to do this?
> ⬇️ Suggested change
>
> -from bcc import BPF, USDT
>
> +# Test will be skipped if we don't have bcc installed
>
> +try:
>
> +    from bcc import BPF, USDT # type: ignore[import]
>
> +except ImportError:
>
> +    pass
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/26531#pullrequestreview-1186573943>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMONZ75LPPNTU5YSG5KPUD3WI7BDRANCNFSM6AAAAAASERWQKM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

glozow reviewed (commented):
utACK d0c6da92c5a4d2a972f884f29dd716da2040fdb2
don't have much experience with bcc, so hoping to get another review from @0xB10C. Also pinging @jamesob @dergoegge for re-review?

virtu commented on src/validation.cpp:
```diff
@@ -1461,6 +1470,10 @@ MempoolAcceptResult AcceptToMemoryPool(Chainstate& active_chainstate, const CTra
 
         for (const COutPoint& hashTx : coins_to_uncache)
             active_chainstate.CoinsTip().Uncache(hashTx);
+        TRACE2(mempool, rejected,
+                tx->GetHash().data(),
+                result.m_state.GetRejectReason().c_str()
+        );
```
Last time I checked event names could not be reused, so the package rejection code would have to use a new event.

On the bright side, having separate rejection events might help differentiating between them and avoid conflating regular vs. package-based rejections.

0xB10C commented:
ACK d0c6da92c5a4d2a972f884f29dd716da2040fdb2

Code reviewed the mempool tracepoint interface tests and the `mempool_monitor.py` example script including the BCC code for both. Tracepoint placement also looks good and the tracepoint arguments match whats added in the documentation.

Two things to note:
- the `mempool_monitor.py` demo script will, over time, use more and more system resources (RAM and CPU) as mempool events are added to the `events` list. The list is never cleared. The entries in the `events` list are used to calculate the statistics, so for a large list it takes longer to calculate. I noticed this after running the demo script for >24h. I don't think it's a problem for a *demo* script, though.
- I still think it makes sense to also pass the time a transaction was added to the mempool in the transaction removed tracepoint as noted in https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1045624376. This allows calculating the time the transaction was in the mempool before being mined/evicted. (I think this is an oversight as you addressed this similar comment https://github.com/bitcoin/bitcoin/pull/26531#discussion_r1045624415 right after it).

I've also setup a quick-and-dirty dashboard using these tracepoints on https://bitcoind.observer/d/jRIa_aaVk/mempool (might turn this off at somepoint again).

achow101 commented:
ACK 4b7aec2951fe4595946cdc804b0dec1921d79d05

russeree commented:
https://github.com/NYDIG/bitcoin-task-bounty/blob/main/bounties/add-bpf-tracepoints-for-the-mempool.md

This might be of interest. For reference I am no longer and have not been working on my mempool tracepoints PR and I closed it months ago.

xJCPMSx commented:
Top
